File: Data Analyst in R/Step 3 - Data Cleaning in R/2. Data Cleaning in R Advanced/2. Advanced Regular Expressions.r
## 1. Introduction ##

library(stringr)
library(readr)

hn  <-  read_csv("hacker_news.csv")
titles  <-  hn$title
sql_pattern  <-  "(?i)SQL"
sql_counts  <-  sum(str_detect(titles, sql_pattern))

## 2. Capture Groups ##

library(dplyr)
library(stringr)
pattern  <-  "(?i)(python [\\d])"
python_flavors  <-  str_match(titles, pattern)[,2]
python_titles_freq  <-  table(str_to_lower(python_flavors))

## 3. Using Capture Groups to Extract Data ##

python_titles <- titles[str_detect(titles,"[Pp]ython [\\d\\.]+")]
pattern <- "[Pp]ython ([\\d\\.]+)"
python_versions <- str_match(python_titles, pattern)[,2]
python_versions_freq <- table(python_versions)

## 4. Counting Mentions of the "C" Language ##

first_10_matches <- function (data, pattern) {
    matches <- str_detect(data, pattern) # finding pattern matches
    matched_df <- data[matches] # subsetting data (keep only matches)
    head(matched_df, 10) # taking the first ten matched elements
}

# pattern  <-  "\\b[Cc]\\b"
pattern  <-  "\\b[Cc]\\b[^.+]"
first_ten  <-  first_10_matches(titles, pattern)

## 5. Using Lookarounds to Control Matches Based on Surrounding Text ##

pattern  <-  "(?<!Series\\s)\\b[Cc]\\b(?![\\+\\.])"
c_mentions  <-  sum(str_detect(titles, pattern))

## 6. BackReferences: Using Capture Groups in a RegEx Pattern ##

pattern  <-  "\\b(\\w+)\\s\\1\\b"

repeated_words  <-  titles[str_detect(titles, pattern)]

## 7. Challenge: Cleaning our dataset ##

library(dplyr)
library(stringr)
pattern  <-  "(?i)(\\w+SQL)"
hn_sql <-  hn %>% 
    filter(str_detect(title, pattern)) %>%
    mutate(flavor = str_match(title, pattern)[,2]) %>%
    mutate(flavor = str_to_lower(flavor))

hn_sql_flavor_avg <- hn_sql %>%
    select(flavor, num_comments) %>%
    group_by(flavor) %>%
    summarise(avg = mean(num_comments))

## 8. Substituting Regular Expression Matches ##

email_variations  <-  c('email', 'Email', 'e Mail',
                        'e mail', 'E-mail', 'e-mail',
                        'eMail', 'E-Mail', 'EMAIL')
pattern  <-  "(?i)e[\\-\\s]?mail"
email_uniform  <-  str_replace(email_variations, pattern, "email")
titles_clean  <-  str_replace(titles, pattern, "email")

## 9. Extracting Domains from URLs ##

test_urls  <-  c(
 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
 'http://www.interactivedynamicvideo.com/',
 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
 'http://evonomics.com/advertising-cannot-maintain-internet-heres-solution/',
 'HTTPS://github.com/keppel/pinn',
 'Http://phys.org/news/2015-09-scale-solar-youve.html',
 'https://iot.seeed.cc',
 'http://www.bfilipek.com/2016/04/custom-deleters-for-c-smart-pointers.html',
 'http://beta.crowdfireapp.com/?beta=agnipath',
 'https://www.valid.ly'
)
pattern  <-  "(?i)https?://([\\w\\.]+)"

test_urls_clean  <-  str_match(test_urls, pattern)[,2]
domains  <- str_match(hn$url, pattern)[,2]
top_domains  <-  head(table(domains),20)

## 10. Extracting URL Parts Using Multiple Capture Groups ##

# `test_urls` is available from the previous screen
pattern  <-  "(.+)://([\\w\\.]+)/?(.*)"

test_url_parts  <-  str_match(test_urls, pattern)

hn_urls  <-  hn %>%
    mutate(protocol = str_match(url, pattern)[,2]) %>%
    mutate(domain = str_match(url, pattern)[,3]) %>%
    mutate(page_path = str_match(url, pattern)[,4])

File: Data Analyst in Python/Step 2 - Intermediate Python and Pandas/5. Data Cleaning in Python Advanced/2. Advanced Regular Expressions.py
## 1. Introduction ##

import pandas as pd
import re

hn = pd.read_csv("hacker_news.csv")
titles = hn['title']
sql_pattern = r"SQL"
sql_counts = titles.str.contains(sql_pattern, flags=re.I).sum()

## 2. Capture Groups ##

hn_sql = hn[hn['title'].str.contains(r"\w+SQL", flags=re.I)].copy()
hn_sql["flavor"] = hn_sql["title"].str.extract(r"(\w+SQL)", re.I)
hn_sql["flavor"] = hn_sql["flavor"].str.lower()
sql_pivot = hn_sql.pivot_table(index="flavor",values="num_comments", aggfunc='mean')

## 3. Using Capture Groups to Extract Data ##

pattern = r"[Pp]ython ([\d\.]+)"

py_versions = titles.str.extract(pattern)
py_versions_freq = dict(py_versions.value_counts())

## 4. Counting Mentions of the 'C' Language ##

def first_10_matches(pattern):
    """
    Return the first 10 story titles that match
    the provided regular expression
    """
    all_matches = titles[titles.str.contains(pattern)]
    first_10 = all_matches.head(10)
    return first_10

# pattern = r"\b[Cc]\b"
pattern = r"\b[Cc]\b[^.+]"
first_ten = first_10_matches(pattern)

## 5. Using Lookarounds to Control Matches Based on Surrounding Text ##

pattern = r"(?<!Series\s)\b[Cc]\b(?![\+\.])"
c_mentions = titles.str.contains(pattern).sum()

## 6. BackReferences: Using Capture Groups in a RegEx Pattern ##

pattern = r"\b(\w+)\s\1\b"

repeated_words = titles[titles.str.contains(pattern)]

## 7. Substituting Regular Expression Matches ##

email_variations = pd.Series(['email', 'Email', 'e Mail',
                        'e mail', 'E-mail', 'e-mail',
                        'eMail', 'E-Mail', 'EMAIL'])
pattern = r"\be[-\s]?mail"
email_uniform = email_variations.str.replace(pattern, "email", flags=re.I)
titles_clean = titles.str.replace(pattern, "email", flags=re.I)

## 8. Extracting Domains from URLs ##

test_urls = pd.Series([
 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
 'http://www.interactivedynamicvideo.com/',
 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
 'http://evonomics.com/advertising-cannot-maintain-internet-heres-solution/',
 'HTTPS://github.com/keppel/pinn',
 'Http://phys.org/news/2015-09-scale-solar-youve.html',
 'https://iot.seeed.cc',
 'http://www.bfilipek.com/2016/04/custom-deleters-for-c-smart-pointers.html',
 'http://beta.crowdfireapp.com/?beta=agnipath',
 'https://www.valid.ly?param',
 'http://css-cursor.techstream.org'
])
pattern = r"https?://([\w\-\.]+)"

test_urls_clean = test_urls.str.extract(pattern, flags=re.I)
domains = hn['url'].str.extract(pattern, flags=re.I)
top_domains = domains.value_counts().head(5)

## 9. Extracting URL Parts Using Multiple Capture Groups ##

# `test_urls` is available from the previous screen
pattern = r"(https?)://([\w\.\-]+)/?(.*)"

test_url_parts = test_urls.str.extract(pattern, flags=re.I)
url_parts = hn['url'].str.extract(pattern, flags=re.I)

## 10. Using Named Capture Groups to Extract Data ##

# pattern = r"(https?)://([\w\.\-]+)/?(.*)"
pattern = r"(?P<protocol>https?)://(?P<domain>[\w\.\-]+)/?(?P<path>.*)"
url_parts = hn['url'].str.extract(pattern, flags=re.I)